### ISSUE TITLE
[RFC]: DFlash Training Best Practices — recommended configuration and reasoning

### ISSUE THREAD

**Issue by shubhra** [vllm-project/speculators#979]
## Motivation

We ran a series of DFlash training experiments on Qwen3-8B, covering vocabulary size, loss weighting, optimizer choice, anchor count, effective batch size, training data composition, and training schedule. This RFC proposes adopting the resulting configuration as a documented recommended default / best-practices reference for DFlash training in this repo, and surfaces the open questions we were not able to resolve, for visibility and discussion before finalizing anything.

If there's agreement on the substance, the concrete next step would be a PR adding this as a docs page (a draft already exists locally at `docs/user_guide/tutorials/dflash_training_best_practices.md`, linked from `getting_started.md` and `tutorials/index.md`) with the supporting charts included.

### Recommended configuration

| Setting | Recommendation |
|---|---|
| Architecture | DFlash draft model, `block_size=16`, 5 draft layers |
| Vocabulary | Full verifier vocabulary — omit the draft-vocab mapping (`--draft-vocab-size`) entirely |
| Loss | Cross-entropy + D-PACE per-position weighting (`--per-position-loss-weight dpace --dpace-alpha 0.5`) |
| Attention | Causal sliding-window attention (SWA), `--sliding-window 2048`, applied to all draft layers |
| Optimizer | Muon for 2D weight matrices (`--muon-lr` tuned per architecture, ≈3e-3 in our setup) + AdamW for norms/embeddings (`--lr` ≈6e-4) |
| Max anchors | As high as GPU memory allows for your architecture — see "Anchor count" below |
| Effective batch size | Large (hundreds of thousands of tokens/step) — see "Effective batch size" below |
| Data | A large, diverse instruction set with responses regenerated by a larger, same-model-family teacher, with targeted additions blended in from the start of training |
| Schedule | Run the full intended epoch budget as a single continuous job — see "Training schedule" below |

### Reference command

Configuration that produced our best result to date (4x H100, 80GB): full vocabulary, D-PACE, Muon, `max_anchors=1536`, 20 total epochs. Adjust `--max-anchors`, `--total-seq-len`, `--muon-lr`, and `--scheduler-total-steps` for different hardware/dataset sizes (see the relevant sections below before reusing as-is).

\`\`\`bash
torchrun --standalone --nproc-per-node 4 scripts/train.py \\
  --verifier-name-or-path Qwen/Qwen3-8B \\
  --data-path <path-to-prepared-data> \\
  --hidden-states-path <path-to-hidden-states> \\
  --vllm-endpoint http://localhost:<port>/v1 \\
  --save-path <checkpoint-output-dir> \\
  --speculator-type dflash \\
  --block-size 16 \\
  --num-layers 5 \\
  --target-layer-ids 2 10 18 26 34 \\
  --sliding-window 2048 \\
  --max-anchors 1536 \\
  --total-seq-len 65536 \\
  --epochs 20 \\
  --scheduler-type cosine \\
  --scheduler-total-steps <steps-per-epoch * 20> \\
  --loss-fn ce \\
  --per-position-loss-weight dpace \\
  --dpace-alpha 0.5 \\
  --optimizer muon \\
  --lr 6e-4 \\
  --muon-lr 3e-3 \\
  --muon-momentum 0.95 \\
  --muon-weight-decay 0.1 \\
  --muon-ns-steps 5 \\
  --muon-adjust-lr-fn match_rms_adamw \\
  --checkpoint-freq 0.1
\`\`\`

`--draft-vocab-size`, `--token-freq-path`, `--d2t-path`, and `--t2d-path` are intentionally omitted (see "Vocabulary" below).

### Vocabulary

The common EAGLE-style approach of shrinking the draft model's vocabulary (a `d2t`/`t2d` mapping down to, e.g., 32,000 tokens) simplifies per-token prediction, but reduces accuracy on rarer tokens and adds a mapping step. DFlash drafts an entire block of tokens in a single parallel forward pass rather than one token at a time, and is correspondingly less sensitive to vocabulary size than autoregressive drafters. In our tests, switching from a 32,000-token draft vocabulary to the full 151,936-token verifier vocabulary increased average acceptance length by approximately +0.3 tokens/round, with no additional inference cost.

### Loss weighting

DFlash's default loss applies a fixed exponential decay (`gamma`) to de-weight later positions in the block. D-PACE instead adapts the per-position weighting dynamically over training rather than fixing it in advance. Across every architecture variant tested (different anchor counts, different optimizers), D-PACE matched or exceeded the fixed-gamma baseline. We did not find a strong reason to move `--dpace-alpha` away from 0.5.

### Optimizer

Muon (2D weight matrices) + AdamW (norms, biases, embeddings, LM head) performed as well as or better than AdamW alone across every configuration tested.

**LR search methodology.** A short ablation on a small data subset can mislead: the LR schedule's warmup/decay shape has a significant effect, so `--scheduler-total-steps` in an ablation run should be pinned to the step count of the intended full production run. We identified `--muon-lr 3e-3` via a two-stage search (wide grid, then a narrower sweep around the best region) at our specific architecture (1,536 anchors, 65,536-token sequences).

**The `--lr`/`--muon-lr` coupling.** If `--muon-lr` is not passed explicitly, it defaults to `10 * --lr` (`scripts/train.py:1308-1309`). Sweeping `--lr` alone therefore sweeps both LRs together at a fixed 10:1 ratio — a reasonable one-dimensional simplification, but it never tests whether 10:1 is the correct *ratio*. We also noticed the effective resolved value isn't surfaced anywhere at runtime (not logged, not in `train_command.txt`), which made it hard to later confirm whether a given run's Muon LR had been independently tuned or was riding the default — worth a small follow-up fix regardless of the rest of this RFC.

### Anchor count

Higher `--max-anchors` consistently improved acceptance length in a controlled, single-variable ablation. The practical ceiling is determined by GPU memory and interacts with `--total-seq-len`:

| GPU | Max anchors | Sequence length |
|---|---|---|
| H100 (80GB) | 1,536 | 65,536 tokens |
| H200 | 2,048 | 147,456 tokens |

Exceeding this ceiling produces an OOM either immediately or partway through a full epoch once real batch-content variance is encountered (short ablation runs underestimate this). Leave headroom below the first observed OOM point.

**Open caveat.** A full training run at 2,048 anchors did not outperform a comparable run at 1,536 anchors on final acceptance length, despite the anchor-count ablation predicting the opposite — and this was not from a skipped LR search; the 2,048-anchor run's LR went through the same two-stage ablation process and converged cleanly. This is unresolved, not attributable to a known cause. Candidate explanations include the fixed 10:1 Muon:AdamW ratio above, or an anchor-count/architecture interaction not yet isolated.

### Effective batch size

Effective batch size (GPU count × sequence length, tokens/step) was substantially larger in our best configuration than typical LLM fine-tuning setups — ~262,144 tokens/step (4x H100, 65,536-token sequences). Two reasons to keep this large:

1. D-PACE applies a distinct per-position loss term at each of `block_size` positions, and `--max-anchors` samples many anchor positions per sequence; a larger effective batch aggregates more of these terms per gradient update, reducing gradient variance.
2. Muon's orthogonalization step is sensitive to gradient noise; larger batches produced more stable convergence in our runs, consistent with published Muon guidance.

### Training data

- Start from a broad instruction dataset with **responses regenerated by a larger teacher model from the same model family as the target** (we used Qwen3-8B as target, Qwen3-235B-A22B-Instruct-2507 as teacher), rather than the dataset's original responses or the target's own outputs. We did not directly test self-distillation or an unrelated-family teacher against this.
- When a specific weakness is identified in evaluation, source additional targeted prompts and **blend the fresh responses in from the start of training**, not only via a later-epoch extension. Blending from epoch 0 gave a consistent +0.1–0.15 avg acceptance length improvement in our tests.
- An FP8 teacher gave ~1.7x faster generation than BF16 with no detectable quality difference, given proper tensor-parallel sizing.

### Training schedule

More total epochs consistently improved results; we never observed a point where additional training clearly hurt. We repeatedly staged training in two phases (N epochs, evaluate, resume for N more with the schedule re-pinned) — every time, the first checkpoint after resuming showed a measurable regression before recovering, a re-annealing artifact rather than evidence the extension is unhelpful. No plateau was observed through 20 total epochs in our best run. **Recommendation: run the full intended epoch budget as a single continuous job** rather than staged resumes, to avoid the dip-and-recovery cycle.

### Results

Incremental application of this recipe on our internal benchmark suite (RedHatAI `speculator_benchmarks`, 9 categories, avg acceptance length):

| Configuration | Avg. acceptance length |
|---|---|
| Reduced (32k) vocabulary baseline | 3.425 |
| Full vocabulary, initial pass (prior to anchor/optimizer tuning) | 3.206 |
| + Muon, tuned anchor count | 3.612 |
| + targeted code/multilingual data, 20 epochs | 3.720 |
| Fresh restart on the same recipe and data, 20 epochs (current best) | **3.737** |

The improvement is consistent per-category and generalizes to GPQA Diamond, LiveCodeBench, and a separate suite (NVIDIA SPEED-Bench, 11 categories) — charts available in the linked docs draft.

## Any Other Things

Known, unresolved gaps rather than settled conclusions:

- **Code generation** is the most consistent weak spot across every benchmark tested, including two independent third-party code benchmarks. More/better code-specific data is the leading hypothesis, not yet confirmed.
- **Translation** performance lags other categories on our suite, though the absolute gap is small; a broader multilingual pass (not narrowly translation-focused) seems more promising than over-indexing on one benchmark's translation subset.
- Anchor-count ceilings were not cross-validated across GPU generations beyond H100/H200.
- **The anchor-count result above is unresolved, not merely under-tested.** Two full runs, each with an independently and properly ablated LR, still showed the lower-anchor configuration on top. A follow-up varying the Muon:AdamW LR ratio directly (not just their shared scale) at the higher anchor count seems like a reasonable next step before concluding more than "higher anchor counts helped in one isolated, single-variable test."
- Separately, we noticed `_resolve_scheduler_steps` (`src/speculators/train/trainer.py:128-138`) already auto-derives `scheduler_total_steps` from the real packed-batch count via `len(train_loader)`, which we were unaware of and never relied on — we manually recalibrated it from measured token counts on every run instead. If this default is reliable for online training with resumed checkpoints, it's worth documenting more prominently.

**Comment by speedyshreya** [comment]
Hi, I'd like to work on this if available, could you please assign it to me?

**Comment by shubhra** [comment]
@speedyshreya - a PR that addresses these changes: https://github.com/vllm-project/speculators/pull/980 has already been merged last week. Thanks!
